fix(amf): Crash observed when DNN Length is incorrectly encoded by UE

sreedharkumartn opened the pull request:
Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary
Fixing the issue #12279.

**ISSUE:**
- DNN IE : DNN Content Length + DNN Contents
- DNN Content Length = sizeof(unsigned char) + strlen("DNN Contents")
where sizeof(unsigned char) stores the length of DNN.
If somehow this value is not encoded correctly by UE then memcpy will crash. Attaching the PCAP snapshot of wrong and correct DNN.

Wrong DNN:
![image](https://user-images.githubusercontent.com/89978170/160392363-147d6f43-1d02-4321-83e9-e00ba45aad5a.png)

Correct DNN:
![image](https://user-images.githubusercontent.com/89978170/160392533-cf434954-9f28-4da1-a1a4-830edc76f17a.png)

**FIX:**
- Added the additional check for DNN Length.

<!-- Enumerate changes you made and why you made them -->

## Test Plan
- Added New Unit test case and validated with UT.
![UT 1](https://user-images.githubusercontent.com/89978170/160392912-375830c5-16da-482c-925f-92d23e55542a.jpg)

- Basic sanity with UERANSIM.
Pcap snap:
![image](https://user-images.githubusercontent.com/89978170/160393191-2a851c59-00ba-428d-887f-4fb28bd73ad0.png)

mme log:
[mme.log](https://github.com/magma/magma/files/8362411/mme.log)


<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

sreedharkumartn committed:
fix(amf): Crash observed when DNN Length is incorrectly encoded by UE

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

ganeshg87 reviewed (approved):
lgtm

rsarwad commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -915,6 +915,25 @@ TEST(test_optional_dnn_pdu, test_pdu_session_establish_optional) {
+
+  uint32_t len = sizeof(pdu) / sizeof(uint8_t);
+
+  NAS5GPktSnapShot nas5g_pkt_snap;
```
Remove unused variable, nas5g_pkt_snap

sreedharkumartn replied:
Addressed Review Comment

rsarwad commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -915,6 +915,25 @@ TEST(test_optional_dnn_pdu, test_pdu_session_establish_optional) {
+                     0x22, 0x04, 0x01, 0x00, 0x00, 0x01, 0x25, 0x08,
+                     0x08, 0x64, 0x65, 0x66, 0x61, 0x75, 0x6c, 0x74};
+
+  uint32_t len = sizeof(pdu) / sizeof(uint8_t);
```
Result of sizeof(uint8_t) is always one, right? 
Why do we need this statement?

sreedharkumartn replied:
Addressed Review Comment

rsarwad commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -915,6 +915,25 @@ TEST(test_optional_dnn_pdu, test_pdu_session_establish_optional) {
+  uint32_t len = sizeof(pdu) / sizeof(uint8_t);
+
+  NAS5GPktSnapShot nas5g_pkt_snap;
+  ULNASTransportMsg pdu_sess_est_req;
```
memset is not required, instead you can initialize all parameters to "zero"

sreedharkumartn replied:
Addressed Review Comment

rsarwad commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -915,6 +915,25 @@ TEST(test_optional_dnn_pdu, test_pdu_session_establish_optional) {
+  bool decode_res = false;
+  memset(&pdu_sess_est_req, 0, sizeof(ULNASTransportMsg));
+
+  decode_res = decode_ul_nas_transport_msg(&pdu_sess_est_req, pdu, len);
```
No need of decode_res. It can coded as 
EXPECT_EQ((decode_ul_nas_transport_msg(&pdu_sess_est_req, pdu, len)), false);

sreedharkumartn replied:
Addressed Review Comment

rsarwad commented on lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp:
```diff
@@ -44,6 +44,14 @@ int DNNMsg::DecodeDNNMsg(DNNMsg* dnn_message, uint8_t iei, uint8_t* buffer,
+
+  if (ielen <= dnn_len) {
+    OAILOG_ERROR(LOG_NAS5G,
+                 "Mismatch Length: IE length : %d, DNN String length: %d",
```
Both variables are of type, unit8_t. Change the format specifier to %u

sreedharkumartn replied:
Addressed Review Comment

rsarwad commented on lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GDNN.cpp:
```diff
@@ -44,6 +44,14 @@ int DNNMsg::DecodeDNNMsg(DNNMsg* dnn_message, uint8_t iei, uint8_t* buffer,
+
+  if (ielen <= dnn_len) {
+    OAILOG_ERROR(LOG_NAS5G,
+                 "Mismatch Length: IE length : %d, DNN String length: %d",
```
Minimum length of DNN is 3 and maximum length of DNN is 102. Can we add this check also?

sreedharkumartn replied:
Addressed Review Comment

sreedharkumartn committed:
Addressed Review Comments (R1)

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

ardzoht reviewed (commented):
not related to the PR, but can we update `dnn_len` on line 61 as there is an outer variable with the same name? this is a legit warning by cpplint

sreedharkumartn committed:
Addressed Review Comment (R2)

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

rsarwad commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -915,6 +915,20 @@ TEST(test_optional_dnn_pdu, test_pdu_session_establish_optional) {
+                     0x01, 0xc1, 0xff, 0xff, 0x91, 0x12, 0x0a, 0x81,
+                     0x22, 0x04, 0x01, 0x00, 0x00, 0x01, 0x25, 0x08,
+                     0x08, 0x64, 0x65, 0x66, 0x61, 0x75, 0x6c, 0x74};
+
```
No need of of this line

sreedharkumartn replied:
Addressed Review Comment

rsarwad commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -915,6 +915,20 @@ TEST(test_optional_dnn_pdu, test_pdu_session_establish_optional) {
+                     0x08, 0x64, 0x65, 0x66, 0x61, 0x75, 0x6c, 0x74};
+
+  uint32_t len = sizeof(pdu);
+
```
No need of of this line

sreedharkumartn replied:
Addressed Review Comment

sreedharkumartn commented:
> not related to the PR, but can we update `dnn_len` on line 61 as there is an outer variable with the same name? this is a legit warning by cpplint

Addressed Review Comment.

sreedharkumartn committed:
Addressed Review Comments (R3)

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

rsarwad reviewed (approved):
LGTM
